apollo-router/src/plugins/traffic_shaping/rate/service.rs

```diff
@@ -32,25 +37,43 @@ where
     type Future = ResponseFuture<S::Future>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        let mut window_start = self.window_start.read().unwrap().elapsed();
-        let time_unit = self.rate.per();
+        let time_unit = self.rate.per().as_millis() as u64;
 
-        if window_start > time_unit {
-            let new_window_start = Instant::now();
-            *self.window_start.write().unwrap() = new_window_start;
-            window_start = new_window_start.elapsed();
+        let updated =
+            self.window_start
+                .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |window_start| {
+                    let duration_start = Duration::from_millis(window_start);
+                    let duration_now = Duration::from_millis(
+                        SystemTime::now()
+                            .duration_since(UNIX_EPOCH)
+                            .expect("system time must be after EPOCH")
+                            .as_millis() as u64,
+                    );
+                    if duration_now.saturating_sub(duration_start) > self.rate.per() {
+                        Some(
+                            SystemTime::now()
+                                .duration_since(UNIX_EPOCH)
+                                .expect("system time must be after EPOCH")
+                                .as_millis() as u64,
+                        )
+                    } else {
+                        None
+                    }
+                });
+        // If it has been updated
+        if let Ok(_updated_window_start) = updated {
```

Comment:
it might still be racy here, ie thread 1 has updated the window but not yet the number of requests, so thread 2 reads the new window_start with the old previous_nb_requests

I'm not sure it will be a big issue though

Reply:
I know but I think it's better than lock these values together. We won't have a hard fixed numbers of requests per interval but the rate error won't be so big I think.